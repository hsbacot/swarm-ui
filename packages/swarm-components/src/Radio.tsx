import * as React from 'react';

interface Props {
	/**
	 * Whether the radio button is checked.
	 */
	checked?: boolean,
	/**
	 * Whether the input should be interactive.
	 */
	disabled?: boolean,
	/**
	 * A unique identifier for the input.
	 */
	id?: string,
	/**
	 * Used to associate a group of radio buttons.
	 * Only one radio button in a group can be selected.
	 */
	name: string,
	/**
	 * Value of the input.
	 */
	value: string,
	/**
	 * String label for the input. Will override any children values.
	 */
	label?: string,
	/**
	 * Use children as an alternative to the `label` prop for more complex input labels.
	 * The `label` prop will override children.
	 */
	children?: React.ReactElement,
};

// Can not inline css vars as color to Icon. Gray 5 icon fill on disbaled
const Radio: React.FC<Props> = (props) => {
	const { checked, label, id, disabled, value, children, name, ...other } = props;

	return (
		<label data-swarm-radio={disabled ? 'disabled' : 'default'}>
			<span
				data-swarm-radio-field={checked ? 'checked' : 'unchecked'}
				tabIndex={0}
				role="checkbox"
				aria-checked={checked}
			>
				{checked && <span data-swarm-radio-dot />}
			</span>
			<input
				type="radio"
				id={id}
				name={name}
				checked={checked}
				disabled={disabled}
				value={value}
				{...other}
			/>
			<span>{label || children}</span>
		</label>
	);
};

export { Radio };
