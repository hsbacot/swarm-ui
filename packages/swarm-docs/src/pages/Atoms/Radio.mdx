import { Radio } from '@meetup/swarm-components';
import Example, { GroupRadioButtonsExample } from '../../components/examples/radio';
import PropsTable from '../../components/propstable';

# Radio
---
A wrapper for an html `<input type="radio">`.

<br/>
<br/>

### Snippet
---

Use this interactive snippet to try the component:
<br/>

```.jsx
<Radio name="swarm" label="swarm" value="1" checked={true} onChange={(e) => alert(e.target.value)} />
```

<br/>
<br/>

### Props
---
<br/>
<PropsTable propMetaData={Radio.__docgenInfo} />

<br/>
<br/>

### a11y
---
<br/>

Radio buttons should be used when only one option from a group can be selected.

<br/>

**Keyboard interactions:**
* **↑/↓** or **←/→**    ... to select an option.
* **Tab**               ... to move to the next element.

<br/>
<br/>

### Examples
---
<br/>

**These are examples of single (not-grouped) Radio buttons:**
<br/>

<Example />

<br/>

**This is an example of grouped radio buttons and how to manage state:**
<br/>

<GroupRadioButtonsExample />
