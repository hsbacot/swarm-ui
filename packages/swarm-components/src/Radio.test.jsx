import React from 'react';
import { shallow } from 'enzyme';
import { Radio } from './Radio';

describe('Radio', () => {
	const mockOnChange = jest.fn();
	const testCases = [
		['Default', <Radio name="radio" key="0" label="Default" />],
		[
			'Children',
			<Radio name="radio" key="1">
				Children
			</Radio>,
		],
		['Checked', <Radio name="radio" key="2" label="Checked" checked />],
		['Disabled', <Radio name="radio" key="3" label="Disabled" disabled />],
		[
			'Checked Disabled',
			<Radio name="radio" key="4" label="Checked Disabled" checked disabled />,
		],
		['With Id', <Radio name="radio" id="radio" key="5" label="Id" />],
		[
			'With onChange',
			<Radio name="radio" key="6" label="onChange" onChange={mockOnChange} />,
		],
	];

	test.each(testCases)('Snapshot: %s', (description, element) => {
		expect(shallow(element)).toMatchSnapshot();
	});
});
