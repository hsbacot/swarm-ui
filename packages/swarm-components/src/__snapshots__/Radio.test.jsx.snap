// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Snapshot: Checked 1`] = `
<label
  data-swarm-radio="default"
>
  <span
    aria-checked={true}
    data-swarm-radio-field="checked"
    role="checkbox"
    tabIndex={0}
  >
    <span
      data-swarm-radio-dot={true}
    />
  </span>
  <input
    checked={true}
    name="radio"
    type="radio"
  />
  <span>
    Checked
  </span>
</label>
`;

exports[`Radio Snapshot: Checked Disabled 1`] = `
<label
  data-swarm-radio="disabled"
>
  <span
    aria-checked={true}
    data-swarm-radio-field="checked"
    role="checkbox"
    tabIndex={0}
  >
    <span
      data-swarm-radio-dot={true}
    />
  </span>
  <input
    checked={true}
    disabled={true}
    name="radio"
    type="radio"
  />
  <span>
    Checked Disabled
  </span>
</label>
`;

exports[`Radio Snapshot: Children 1`] = `
<label
  data-swarm-radio="default"
>
  <span
    data-swarm-radio-field="unchecked"
    role="checkbox"
    tabIndex={0}
  />
  <input
    name="radio"
    type="radio"
  />
  <span>
    Children
  </span>
</label>
`;

exports[`Radio Snapshot: Default 1`] = `
<label
  data-swarm-radio="default"
>
  <span
    data-swarm-radio-field="unchecked"
    role="checkbox"
    tabIndex={0}
  />
  <input
    name="radio"
    type="radio"
  />
  <span>
    Default
  </span>
</label>
`;

exports[`Radio Snapshot: Disabled 1`] = `
<label
  data-swarm-radio="disabled"
>
  <span
    data-swarm-radio-field="unchecked"
    role="checkbox"
    tabIndex={0}
  />
  <input
    disabled={true}
    name="radio"
    type="radio"
  />
  <span>
    Disabled
  </span>
</label>
`;

exports[`Radio Snapshot: With Id 1`] = `
<label
  data-swarm-radio="default"
>
  <span
    data-swarm-radio-field="unchecked"
    role="checkbox"
    tabIndex={0}
  />
  <input
    id="radio"
    name="radio"
    type="radio"
  />
  <span>
    Id
  </span>
</label>
`;

exports[`Radio Snapshot: With onChange 1`] = `
<label
  data-swarm-radio="default"
>
  <span
    data-swarm-radio-field="unchecked"
    role="checkbox"
    tabIndex={0}
  />
  <input
    name="radio"
    onChange={[MockFunction]}
    type="radio"
  />
  <span>
    onChange
  </span>
</label>
`;
